cmd/simulator/metrics/metrics.go

```diff
@@ -40,3 +57,77 @@ func NewMetrics(reg prometheus.Registerer) *Metrics {
 	reg.MustRegister(m.IssuanceToConfirmationTxTimes)
 	return m
 }
+
+type MetricsServer struct {
+	metricsPort     string
+	metricsEndpoint string
+
+	cancel   context.CancelFunc
+	stopOnce sync.Once
+	stopCh   chan struct{}
+}
+
+func (m *Metrics) Serve(ctx context.Context, metricsPort string, metricsEndpoint string) *MetricsServer {
+	ctx, cancel := context.WithCancel(ctx)
+	// Create a prometheus server to expose individual tx metrics
+	server := &http.Server{
+		Addr: fmt.Sprintf(":%s", metricsPort),
+	}
+
+	// Start up go routine to listen for SIGINT notifications to gracefully shut down server
+	go func() {
+		// Blocks until signal is received
+		<-ctx.Done()
+
+		if err := server.Shutdown(ctx); err != nil {
+			log.Error("Metrics server error: %v", err)
+		}
+		log.Info("Received a SIGINT signal: Gracefully shutting down metrics server")
+	}()
+
+	// Start metrics server
+	ms := &MetricsServer{
+		metricsPort:     metricsPort,
+		metricsEndpoint: metricsEndpoint,
+		stopCh:          make(chan struct{}),
+		cancel:          cancel,
+	}
+	go func() {
+		defer close(ms.stopCh)
+
+		http.Handle(metricsEndpoint, promhttp.HandlerFor(m.reg, promhttp.HandlerOpts{Registry: m.reg}))
+		log.Info(fmt.Sprintf("Metrics Server: localhost:%s%s", metricsPort, metricsEndpoint))
+		if err := server.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+			log.Error("Metrics server error: %v", err)
+		}
+	}()
+
+	return ms
+}
+
+func (ms *MetricsServer) Shutdown() {
+	ms.stopOnce.Do(func() {
```

Comment:
nit: I don't think stopOnce is needed since calling cancel and reading from stopCh can both be repeated.

I think we can simplify the shutdown, but no need to block the PR for this.

Reply:
Will remove